staking-pool-factory/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_sdk::collections::Set;
+use near_sdk::json_types::{Base58PublicKey, U128};
+use near_sdk::{env, ext_contract, near_bindgen, AccountId, Balance, Promise, PromiseOrValue};
+use serde::{Deserialize, Serialize};
+
+mod utils;
+use crate::utils::*;
+
+/// The 30 NEAR tokens required for the storage of the staking pool.
+const MIN_ATTACHED_BALANCE: Balance = 30_000_000_000_000_000_000_000_000;
+
+pub mod gas {
+    use near_sdk::Gas;
+
+    /// The base amount of gas for a regular execution.
+    const BASE: Gas = 20_000_000_000_000;
+
+    /// The amount of Gas the contract will attach to the promise to create the staking pool.
+    /// The base for the execution and the base for staking action to verify the staking key.
+    pub const STAKING_POOL_NEW: Gas = BASE * 2;
+
+    /// The amount of Gas the contract will attach to the callback to itself.
+    /// The base for the execution and the base for whitelist call or cash rollback.
+    pub const CALLBACK: Gas = BASE * 2;
+
+    /// The amount of Gas the contract will attach to the promise to the whitelist contract.
+    /// The base for the execution.
+    pub const WHITELIST_STAKING_POOL: Gas = BASE;
+}
+
+/// There is no deposit balance attached.
+const NO_DEPOSIT: Balance = 0;
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct StakingPoolFactory {
+    /// Account ID of the staking pool whitelist contract.
+    staking_pool_whitelist_account_id: AccountId,
+
+    /// The account ID of the staking pools created.
+    staking_pool_account_ids: Set<AccountId>,
```

Comment:
Isn't this information stored in the whitelist contract?

Reply:
It's different, here it mostly to avoid creation of the staking pools with the same name.

Reply:
Yeah, but can't you make a promise call to the whitelist to do that?

Reply:
That's extra call, but in theory nothing breaks if we don't maintain the `Set` internally. It just saves some gas if 2 people try to create the same pool. Then one of the creations is going to fail.

Reply:
Okay makes sense to me.  Then everything LGTM.